File: Problem2_DP.java
//Time Complexity : O(m*n)
//Space Complexity : O(m*n)
//Code run successfully on LeetCode.

public class Problem2_DP {
	
	public int change(int amount, int[] coins) {
        
		if(coins.length ==0 || coins == null)
			return 0;
		
		int m = coins.length;
		int n = amount;
		
		int[][] dp = new int[m+1][n+1];
		
		for(int i =0; i< m+1; i++)
			dp[i][0] = 1;
		
		for(int i =1; i < m+1; i++) {
			for(int j =1; j < n+1; j++) {
				
				if(j < coins[i-1])
					dp[i][j] = dp[i-1][j];
				
                else
				   dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]];
			}
		}
		
		return dp[m][n];
    }

}

File: problem2_DP.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        int m {amount+1};
        int n {static_cast<int>(coins.size()+1)};
        vector<vector<int>> arr(n,vector<int> (m));
        arr.at(0).at(0) = 1;

        for(int i{1};i<=coins.size();++i){
            for(int j{0};j<=amount;++j){
                if(j<coins.at(i-1)){
                    arr.at(i).at(j) = arr.at(i-1).at(j);
                }
                else{
                    int val1{arr.at(i).at(j-coins.at(i-1))};
                    int val2{arr.at(i-1).at(j)};
                    arr.at(i).at(j) = val1+val2;
                }
            }
        }

        return arr.at(coins.size()).at(amount);
    }
};
